lib/SyntaxParse/SyntaxTreeCreator.cpp

```diff
@@ -144,15 +144,18 @@ SyntaxTreeCreator::recordRawSyntax(syntax::SyntaxKind kind,
   SmallVector<const RawSyntax *, 16> parts;
   parts.reserve(elements.size());
   size_t TextLength = 0;
+  size_t TotalSubNodeCount = 0;
   for (OpaqueSyntaxNode opaqueN : elements) {
     auto Raw = static_cast<const RawSyntax *>(opaqueN);
     parts.push_back(Raw);
     if (Raw) {
       TextLength += Raw->getTextLength();
+      TotalSubNodeCount += Raw->getTotalSubNodeCount() + 1;
+      Arena->addChildArena(Raw->getArenaPointer());
```

Comment:
I really don't like `SyntaxTreeCreator` knows the implementation detail of `RawSyntax` too much.
Could we make `RawSyntax::RawSyntax()` a template that receives an iterator range, and do everything inside them in a single loop? So that, `SyntaxTreeCreator` will be like:
```cpp
OpaqueSyntaxNode
SyntaxTreeCreator::recordRawSyntax(
    syntax::SyntaxKind kind,
    ArrayRef<OpaqueSyntaxNode> elements) {
  auto children = llvm::map_range(elements, [](const OpaqueSyntaxNode &elt) {
    return static_cast<const RawSyntax *>(elt);
  });
  auto raw = RawSyntax::make(kind, children, SourcePresence::Present, Arena);  
  return static_cast<OpaqueSyntaxNode>(raw);
}

OpaqueSyntaxNode
SyntaxTreeCreator::makeDeferredLayout(
    syntax::SyntaxKind kind, bool IsMissing,
    const MutableArrayRef<ParsedRawSyntaxNode> &parsedChildren) {
  auto children = llvm::map_range(parsedChildren, [](ParsedRawSyntaxNode &node) {
    return node.takeData();
  });
  auto raw = RawSyntax::make(kind, children, SourcePresence::Present, Arena);
  return static_cast<OpaqueSyntaxNode>(raw);
}
```

Reply:
Probably passing both the end iterator and the size is redundant.
```cpp
OpaqueSyntaxNode
SyntaxTreeCreator::recordRawSyntax(
    syntax::SyntaxKind kind,
    ArrayRef<OpaqueSyntaxNode> elements) {
  auto iter = llvm::map_iterator(elements.begin(), [](const auto &elt) {
    return static_cast<const RawSyntax *>(elt);
  });
  auto raw = RawSyntax::make(kind, iter, elements.size(),
                             SourcePresence::Present, Arena);  
  return static_cast<OpaqueSyntaxNode>(raw);
}
```

Reply:
Or just cast the entire `elements` `ArrayRef` to `RawSyntax *` and pass it to `RawSyntax::make`. We don’t actually need the iterator. 

```cpp
ArrayRef<const RawSyntax *> rawSyntaxElements = llvm::makeArrayRef(static_cast<const RawSyntax *>(elements.data()), elements.size());
```

I think I wasn’t able to do that while `RawSyntax` was still ref-counted (and I didn’t think about the iterator). But now that that’s gone, we should be able to do it.

Reply:
But we still need some mapping iterator from `ParsedRawSyntaxNode` to `const RawSyntax *` for `SyntaxTreeCreator::makeDeferredLayout()`